### DISCUSSION TITLE
Is there a nice way to add modules from multiple assemblies without calling AddModule for each one

### DISCUSSION THREAD

**Discussion by JoshCollis** [discord-net/Discord.Net#2360] [Q&A] [answered]
I really want to seperate out each of my interaction modules into seperate projects, for the purpose of being able to deploy them seperately (and maybe eventually load/unload at runtime as well)

I could call AddModuleAsync on each type one by one in my command service, but I'm wondering if there is a nicer way to discover and load modeules.

I currentyly have a seperate assembly with all my modules, and pass that assembly in to AddModulesAsync, but this does not allow me to deploy individual modules seperately as they don't each have their own DLL when deployed

Any guidance would be greatly appreciated!

**Comment by exsersewo** [comment, marked as answer]
Couldn't you get the assembly that the modules reside in then call `AddModulesAsync` on that?

Something akin to:
`commandService.AddModulesAsync(typeof(SuperAwesomeModule).Assembly, services ?? null);`

**Reply by exsersewo** [reply]
This would give you all current assemblies loaded so tying the above with this should give you what you want.
 `AppDomain.CurrentDomain.GetAssemblies()`

**Reply by JoshCollis** [reply]
> This would give you all current assemblies loaded so tying the above with this should give you what you want. `AppDomain.CurrentDomain.GetAssemblies()`

Amazing, so I suppose I just loop through that and run AddModuleAsync on each Assembly? 

Does this seem like a good design to you?

**Reply by exsersewo** [reply]
> Amazing, so I suppose I just loop through that and run AddModuleAsync on each Assembly?

Yeah, you can start to bring in reflection to ensure that the assembly contains anything inheriting a module to ensure you only load the correct ones.

> Does this seem like a good design to you?

Personally imho, splitting each module into its own project can lead to scalability issues. Though if it's something you want to do to allow reloading of modules to have zero downtime, I can see why you would want to do it.

**Comment by JoshCollis** [comment]
Thanks for your reply,

Like i say, I'm currently doing this which is fine, but what I want to do is have each module separated into its own assembly.  That way they each have their own dll and can be deployed independently.

I would lime to avoid writing an add module line for every single module, I just feel like there ought to be a nice way to discover these modules (maybe by namespace?) And add all of the assemblies

On 12 Jun 2022 16:24, exsersewo ***@***.***> wrote:

Couldn't you get the assembly that the modules reside in then call AddModulesAsync on that?

Something akin to:
commandService.AddModulesAsync(type of(SuperAwesomeModule).Assembly, services ?? null);
